Perf: refactor DistributionsByCountry#report using raw SQL query

coalest opened the pull request:
Doesn't resolve any issue.

### Description
One of the pages that was reported to be slow in production is the `distributions_by_country#report` page.

I tried out two approaches:
1. Fixing the N+1's found in the controller action.
2. Rewriting the service logic in raw SQL.

Approach number 2 seems faster, but potentially less maintainable. Approach number 1 still helps dramatically though.

Request times mentioned below are just for comparison purposes (I still just have the seed data locally).

### Type of change

* Internal refactor (behavior should be identical for users)

### How Has This Been Tested?
- Transferred the old service specs to be query specs.
- Tested locally

### Screenshots
Before any changes (~280 ms):

![Screenshot from 2024-12-10 17-24-20](https://github.com/user-attachments/assets/3ef2ec37-ccc0-46e7-ace8-4c5c56b14583)

After fixing the N+1's using by adding associations to `includes` (~120 ms):

![Screenshot from 2024-12-10 17-24-09](https://github.com/user-attachments/assets/05be1913-9d70-4d01-8588-7e33445a5ace)

<details>
  <summary>View diff of what fixing N+1's look like</summary>
  
  ```diff
diff --git a/app/controllers/distributions_by_county_controller.rb b/app/controllers/distributions_by_county_controller.rb
index 715dc725..28ab4fe5 100644
--- a/app/controllers/distributions_by_county_controller.rb
+++ b/app/controllers/distributions_by_county_controller.rb
@@ -4,7 +4,11 @@ class DistributionsByCountyController < ApplicationController

   def report
     setup_date_range_picker
-    distributions = current_organization.distributions.includes(:partner).during(helpers.selected_range)
+    distributions = current_organization
+      .distributions
+      .includes(partner: { profile: {served_areas: [:county] }})
+      .includes(line_items: [:item])
+      .during(helpers.selected_range)
     @breakdown = DistributionByCountyReportService.new.get_breakdown(distributions)
   end
 end
  ```
</details>

After rewriting with raw SQL (~60 ms):

![Screenshot from 2024-12-10 17-24-34](https://github.com/user-attachments/assets/9c5945e7-9829-40ed-beb9-634e0cd41af0)

dorner commented on app/queries/distribution_summary_by_county_query.rb:
```diff
@@ -0,0 +1,78 @@
+class DistributionSummaryByCountyQuery
```
Is there a reason this needs to be an instance instead of just having a class function? It's easier to test, stub and work with.

dorner replied:
Also can we return a data class instead of a hash?

coalest replied:
> Is there a reason this needs to be an instance instead of just having a class function? It's easier to test, stub and work with.

No real reason, current queries in this folder vary (sometimes `call` is an instance method sometimes a class method). 

Changed it to be a class method.

coalest replied:
> Also can we return a data class instead of a hash?

Fixed. Out of curiosity, why do you prefer data objects rather than hashes?

dorner reviewed (changes requested):
Love the approach overall!

dorner reviewed (changes requested):
Overall looks good to me, although I'm a bit concerned at the test changes. Performance-only fixes shouldn't need to change tests - by definition the way you know it didn't break anything is when the exact same tests pass as before. Can you explain why they were changed?

coalest commented:
I didn't touch the related system or request specs for this report/page. But since I moved the core logic from a service to a query object, so I figured some refactoring of the test was ok. I left the actual assertions the same, but I refactored the setup.

The old setup was including setup from elsewhere (`include_examples "distribution_by_county"`), and it creates some records which weren't needed (user/org_admins). I prefer writing specs where you can see all of the setup in the file rather than DRYing it up which tends to complicate things.

dorner commented:
@coalest I'd really appreciate it if for this PR you could leave the tests as is so we can validate the success of the goal of this PR, which is improving performance without changing anything else.

coalest commented:
Turns out the test was flaky because the JOIN clause was incorrect. It was only working at all because the id's ended up matching by accident sometimes.

coalest commented:
I will revert the changes to the spec regardless.

coalest commented:
I made the new query spec as similar to the old service spec as possible. I had to change the lines calling/executing the query, and then linting made a few additional small changes.

Ready for review again.

dorner commented:
Looks good on my end. @cielf want to kick the tires?

cielf commented:
I think I should.   I'm having a busy end-of-week, so it might not make this release.

cielf commented:
Did some light testing.  Seems ok to me.  Merging.
